test_plans/weak_object.asciidoc

```diff
@@ -0,0 +1,167 @@
+
+* Non-empty `weak_object` created by using `weak_object(const SyclObject &SYCLObj)` constructor
+
+== Tests
+
+=== API tests
+
+General checks for all of created `weak_object`. First initialize `ret` using `try_lock()`, then:
+
+* If the weak_object was constructed with an underlying SYCL object `SyclObj`, check that:
+** `ret.value() == SYCLObj`
+** `std::is_same_v<decltype(ret.value(), SyclObject)>`
+** Call `lock` and verify that the value it returns compares equal to `SyclObj`.
+** Call `reset`, and then verify that `expired()` returns `false`.
+
+* If the weak_object was constructed with no underlying SYCL object, check that:
+** `ret.has_value() == false`
+** Call `lock` and verify that it throws `errc::invalid` exception
+
+=== Constructors
+
+Create instances of `weak_object` using: +
+`constexpr weak_object()` +
+`weak_object(const SyclObject &SYCLObj)` +
+`weak_object(const weak_object &Other)` +
+`weak_object(weak_object &&Other)` +
+
+Perform tests described above in the "API" section.
+
+=== Copy and move assignment operators
+
+Use the existing instances of `weak_object` from the previous section, copy or move them into another objects using following operators: +
+`weak_object &operator=(const SyclObject &SYCLObj)` +
+`weak_object &operator=(const weak_object &Other)` +
+`weak_object &operator=(weak_object &&Other)`
+
+Perform tests described above in the "API" section.
+
+=== swap member function
+
+Using the objects created in the "Constructors" section, do the following:
+
+* Create an empty `weak_object` of the same type by using the default constructor.
+
+* Call `swap` on that object, passing the object from the "Constructors" section.
+
+* Perform tests described above in the "API" section.
+
+=== Expiring
+
+* Create local scope with `SyclObject` and assign it to `weak_object` that is outside this scope. Check that `expired()` returns `false` now and `true` after `SyclObject` is destroyed outside the scope
+
+* Check that `expire()` returns `bool` type using `std::is_same_v`
+
+=== owner_before and owner_less
+
+Create two pairs of `weak_object` objects, the first of which encapsulate the same SYCLObj, and the second are empty
+
+* Check that `owner_before(const weak_object &Other)` returns `false` with passing second `weak_object` for both pairs
+
+* CHeck that `owner_before(const SyclObject &Other)`returns `false` with passing referred SyclObj for both pairs
+
+* Check that `owner_less{}(const T &lhs, const T &rhs)` +
+`owner_less{}(const weak_object<T> &lhs, const weak_object<T> &rhs)` +
+`owner_less{}(const T &lhs, const weak_object<T> &rhs)` +
+`owner_less{}(const weak_object<T> &lhs, const T &rhs)` +
+return `false` if `lhs` and `rhs` `weak_object` or `SYCLObj` they refers are both the same or empty
+
+=== ext_oneapi_owner_before
+Create `SyclObject` and `weak_object` which encapsulate it.
+
+* Check that `SyclObject::ext_oneapi_owner_before` returns `false` with passing created `weak_object`
```

Comment:
These last two sections still need work.  For `owner_before`, I think you want to test the following combinations:

* Verify that `owner_before` compares equivalent for two weak objects that both refer to the same underlying SYCL object:
  * Create a SYCL object of type `T`.  Call it `SyclObject`.
  * Create two `weak_object` objects from `SyclObject`.  Call them `w1` and `w2`.
  * Verify that `w1.owner_before(w2) == false` and `w2.owner_before(w1) == false`.
  * Verify that `w1.owner_before(SyclObject) == false`.
* Verify that `owner_before` compares equivalent for two weak objects that are both empty:
  * Create two empty `weak_object` objects by using the default constructor.  Call them `w1` and `w2`.
  * Verify that `w1.owner_before(w2) == false` and `w2.owner_before(w1) == false`.
* Verify that `owner_before` has some order for two weak object that refer to different underlying SYCL objects:
  * Create two SYCL objects of type `T`.  Call them `SyclObject1` and `SyclObject2`.
  * Create a `weak_object` object from `SyclObject1` and another from `SyclObject2`.  Call them `w1` and `w2`.
  * Verify that exactly one of the following is true:
    * `w1.owner_before(w2) == true && w2.owner_before(w1) == false`, or
    * `w1.owner_before(w2) == false && w2.owner_before(w1) == true`.
  * Verify that `w1.owner_before(SyclObject2) == w1.owner_before(w2)`
  * Verify that `w2.owner_before(SyclObject1) == w2.owner_before(w1)`

The test plan should have another similar section for testing `T.sycl_ext_owner_before`.
The plan should also have a similar section for testing `sycl::ext::oneapi::owner_less`.